test_plans/reference_semantics.asciidoc

```diff
@@ -0,0 +1,191 @@
+2. Check that `t0 == t0`.
+3. Check that `std::hash<T>{}(t0) == std::hash<T>{}(t0)`.
+
+==== Transitivity
+
+This test is repeated for the copy constructor and for copy assignment. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Create a copy of `t0`, `t1`.
+3. Create a copy of `t1`, `t2`.
+4. Check that `t0 == t2` and `t2 == t0`.
+5. Check that `std::hash<T>{}(t0) == std::hash<T>{}(t2)`.
+
+=== Shared tests
+
+The tests in this section tests are run from a host application for all host types and kernel types, and from a kernel function for all kernel types.
+
+The tests below assert equality without the equality operator, such that they may also be executed in a kernel function. The "state" denotes the results of calling member functions of an object.
+
+==== Type traits
+
+For class `T`, check that the following statements evaluate to `true`.
+
+- `std::is_copy_constructible_v<T>`.
+- `std::is_copy_assignable_v<T>`
+- `std::is_destructible_v<T>`
+- `std::is_move_constructible_v<T>`
+- `std::is_move_assignable_v<T>`
+
+==== Copy
+
+This test is repeated for the copy constructor and for copy assignment. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Store the state of `t0`.
+3. Create a copy of `t0`, `t1`.
+4. Check that the state of `t1` is the same as the state stored in Step 2.
+
+==== Move
+
+This test is repeated for the move constructor and for move assignment. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Store the state of `t0`.
+3. Move `t0` to `t1`.
+4. Check that the state of `t1` is the same as the state stored in Step 2.
+
+=== Implementation notes
+The tests in this plan replace `common/common_by_reference.h` and its usages (in `kernel_id_common_reference_semantics.cpp`). The test will be implemented in a `common` header, and be executed from implementation files located in the types' respective directories. The tests in `accessor_legacy` and `event` will be rewritten to use this common header.
+
+=== Propagation of mutations
+
+The tests below tests specify a "mutation" of an instance `T`. This means to change an instance such that the change can be detected in a copy of that instance. This is only possible for the following types:
+
+- `accessor<int, 1, access_mode::read_write, target::host_task>` by writing to the accessor (in the host application):
+* mutation: write to the accessor.
+* detection: read from the accessor.
+- `accessor<int, 1, access_mode::read_write, target::device>` (in a kernel function):
+* mutation: write to the accessor.
+* detection: read from the accessor.
+- `buffer<int>` by obtaining an accessor and writing to the accessor (in the host application):
+* mutation: create an accessor, write to the accessor.
+* detection: create an accessor, read from the accessor.
+- `event` by creating an event with multiple dependent events (in the host application):
+* mutation: resolve an event on the wait list.
+* detection: query the size of the wait list.
```

Comment:
I'm  not sure how this will work.  There is no SYCL API to create an event with dependent events.  I suppose you could submit a `single_task` that has dependent events.

However, I'm not sure how you would "resolve an event on the wait list".  There's no way to manually resolve an event.

Reply:
You should be able to do a single task with a  `while !signaled[0] {}` and then use the host to modify the  `signaled` variable. Allow you to trigger an event on the wait list.